src/util/bson/bson.cpp

```diff
@@ -0,0 +1,1164 @@
+            BsonArray* array;
+        };
+    public:
+        ~BsonContainer() noexcept
+        {
+            if (m_type == DOCUMENT) {
+                delete document;
+                document = NULL;
+            } else {
+                delete array;
+                array = NULL;
+            }
+        }
+
+        BsonContainer(const BsonDocument& v) noexcept : document(new BsonDocument(v)) {
+            m_type = DOCUMENT;
+        }
+        BsonContainer(const BsonArray& v) noexcept : array(new BsonArray(v)) {
+            m_type = ARRAY;
+        }
+        BsonContainer(BsonDocument&& v) noexcept :
+            m_type(DOCUMENT), document(new BsonDocument(std::move(v))) {}
+
+        BsonContainer(BsonArray&& v) noexcept : m_type(ARRAY), array(new BsonArray(std::move(v))) {}
+
+        explicit operator const BsonDocument&&() const {
+            REALM_ASSERT(is_document());
+            return std::move(*document);
+        }
+
+        explicit operator const BsonArray&&() const {
+            REALM_ASSERT(is_array());
+            return std::move(*array);
+        }
+
+        BsonContainer(const BsonContainer& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+        }
+
+        BsonContainer& operator=(const BsonContainer& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+            return *this;
+        }
+
+        BsonContainer& operator=(BsonContainer&& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                if (document) delete document;
+                document = v.document;
+                v.document = NULL;
+            } else {
+                if (array) delete array;
+                array = v.array;
+                v.array = NULL;
+            }
+            return *this;
+        }
+
+        BsonContainer(BsonContainer&& v) {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                if (document) delete document;
+                document = v.document;
+                v.document = NULL;
+            } else {
+                if (array) delete array;
+                array = v.array;
+                v.array = NULL;
+            }
+        }
+
+        bool is_array() const { return m_type == ARRAY; }
+        bool is_document() const { return m_type == DOCUMENT; }
+
+        void push_back(std::pair<std::string, Bson> value) {
+            if (m_type == DOCUMENT) {
+                document->operator[](value.first) = value.second;
```

Comment:
personal preference, but I find this to be a bit more readable `(*document)[value.first] ...`